How Clerk Works (#1735)
clerk/clerk-docs: docs/how-clerk-works/overview.mdx

- **Cookie name:** `__session`
- **Contents:** A Clerk-signed JWT containing [a set of default claims](/docs/backend-requests/resources/session-tokens#default-session-claims), which can also be customized to include additional claims via the Clerk dashboard.
- **Domain:** Set on your application's domain directly, scoped strictly so it cannot be shared across subdomains. This is done for security, to prevent a security issue on a different app on a different subdomain from being able to take over your users' accounts. If you need to send the session token value across subdomain boundaries (like from `example.com` to `api.example.com`), you can [put the token in a request header instead](/docs/backend-requests/making/cross-origin).
- **Expiration:** 60 seconds
- **HttpOnly:** No - must be able to be accessed by client-side SDKs
- **SameSite:** Lax

> The logistics of how these cookies are set and managed are handled by Clerk's frontend SDKs, as many of the details can't be handled purely by an API endpoint response. For this reason, we strongly recommend using one of Clerk'said frontend SDKs to implement authentication flows in general.

Reviewer: ```suggestion
The logistics of how these cookies are set and managed are handled by Clerk's frontend SDKs, as many of the details can't be handled purely by an API endpoint response. For this reason, we strongly recommend using one of Clerk's frontend SDKs to implement authentication flows in general.
```

```js
const publishableKey = 'pk_test_ZXhhbXBsZS5hY2NvdW50cy5kZXYk'
const keyWithoutPrefix = publishableKey.replace('pk_test_', '')

atob(keyWithoutPrefix) // => example.accounts.dev$
```

> [!NOTE]
> In previous versions of Clerk, the Frontend API URL was exposed directly rather than being encoded within a Publishable Key. This was a source of confusion for users, so we transitioned to encoding it as base64 and making it a key.

FAPI manages authentication flows on a per-user basis. For instance, it handles flows for signing up a user, retrieving a user's active sessions, creating an organization on behalf of a user, or fetching a user's organization invites. You can find the complete FAPI documentation [here](https://clerk.com/docs/reference/frontend-api).

FAPI _does not_ handle administrative actions that impact multiple users, such as listing all users, banning users, or impersonating a user. These types of tasks are handled by [the backend API](#backend-api).

> Some tasks, such as [signing up a user](https://clerk.com/docs/reference/frontend-api/tag/Sign-Ups#operation/createSignUps), don't require authentication, as that would defeat the purpose of the endpoint. However, endpoints designed for authenticated users, like [updating a user's details](https://clerk.com/docs/reference/frontend-api/tag/User#operation/patchUser), require FAPI to 1. identify the user making the request and 2. verify their authorization. This ensures that users cannot modify another user's details. Typically, this is achieved by sending a signed token with the request, either as a cookie or a header. You can [learn more about Clerk's authentication tokens later in this guide](#stateful-authentication).

Reviewer 2: ```suggestion
Some tasks, such as [signing up a user](https://clerk.com/docs/reference/frontend-api/tag/Sign-Ups#operation/createSignUps), don't require authentication, as that would defeat the purpose of the endpoint. However, endpoints designed for authenticated users, like [updating a user's details](https://clerk.com/docs/reference/frontend-api/tag/User#operation/patchUser), require FAPI to first identify the user making the request and then verify their authorization. This ensures that users cannot modify another user's details. Typically, this is achieved by sending a signed token with the request, either as a cookie or a header. You can [learn more about Clerk's authentication tokens later in this guide](#stateful-authentication).
```

> **Customizability:** You have full control over every part of the authentication flow, including HTML structure, CSS, and the logic/ordering of how the authentication flow works.

## The backend API

The frontend API (FAPI) is designed for use primarily from the frontend of your application. Its methods focus on signing in users and managing user-related resources and data once they are authenticated. However, as an application developer, you might also need to perform administrative tasks, such as modifying multiple user or organization details, retrieving a list of all users, banning or impersonating a user, and more.

To maintain security, these administrative tasks should only be executed on the server side using a secret key inaccessible to your users or the browser. These operations are handled by a separate API known as the backend API (BAPI). You can find the BAPI documentation [here](https://clerk.com/docs/reference/backend-api).

> Like FAPI, while you can interact directly with BAPI, most developers opt to use Clerk's SDKs for smoother integration with their preferred language or framework. Documentation for Clerk's SDKs is available in [the left sidebar of the docs](https://clerk.com/docs).

Reviewer 2: ahh good point - what do you think about this update:

```suggestion
Like FAPI, while you can interact directly with BAPI, most developers opt to use Clerk's SDKs for smoother integration with their preferred language or framework. Documentation for Clerk's SDKs is available in [the left sidebar of the docs](https://clerk.com/docs). That being said, FAPI is an extremely more complex and nuanced API that relies on more custom logic outside of its endpoints to create a functional SDK on top of it. As such, **interacting directly with FAPI is not recommended**, whereas interacting directly with BAPI is generally reasonable.
```

---
title: How Clerk works
description: Learn how Clerk is architected and how it works under the hood.
---

This guide provides a deep dive into Clerk's architecture and internal workings. For developers who are simply looking to add authentication to their app, see the [quickstart guides](/docs/quickstarts/overview).

## The frontend API

> When you create a new application through the [Clerk Dashboard](https://dashboard.clerk.com/), Clerk provisions a dedicated frontend API (FAPI) instance for your application. It is hosted at `https://<slug>.accounts.dev` in development environments, where `<slug>` is a unique identifier generated for your application. You can find your application's FAPI URL in the [**Domains**](https://dashboard.clerk.com/last-active?path=domains) page of the Clerk Dashboard.

Reviewer 2: ```suggestion
When you create a new application through the [Clerk Dashboard](https://dashboard.clerk.com/), Clerk provisions a dedicated frontend API (FAPI) instance for your application. It is hosted at `https://<slug>.clerk.accounts.dev` in development environments, where `<slug>` is a unique identifier generated for your application. You can find your application's FAPI URL in the [**Domains**](https://dashboard.clerk.com/last-active?path=domains) page of the Clerk Dashboard.
```

```js
const publishableKey = 'pk_test_ZXhhbXBsZS5hY2NvdW50cy5kZXYk'
const keyWithoutPrefix = publishableKey.replace('pk_test_', '')

atob(keyWithoutPrefix) // => example.accounts.dev$
```

> [!NOTE]
> In previous versions of Clerk, the Frontend API URL was exposed directly rather than being encoded within a Publishable Key. This was a source of confusion for users, so we transitioned to encoding it as base64 and making it a key.

FAPI manages authentication flows on a per-user basis. For instance, it handles flows for signing up a user, retrieving a user's active sessions, creating an organization on behalf of a user, or fetching a user's organization invites. You can find the complete FAPI documentation [here](https://clerk.com/docs/reference/frontend-api).

FAPI _does not_ handle administrative actions that impact multiple users, such as listing all users, banning users, or impersonating a user. These types of tasks are handled by [the backend API](#backend-api).

> Some tasks, such as [signing up a user](https://clerk.com/docs/reference/frontend-api/tag/Sign-Ups#operation/createSignUps), don't require authentication, as that would defeat the purpose of the endpoint. However, endpoints designed for authenticated users, like [updating a user's details](https://clerk.com/docs/reference/frontend-api/tag/User#operation/patchUser), require FAPI to 1. identify the user making the request and 2. verify their authorization. This ensures that users cannot modify another user's details. Typically, this is achieved by sending a signed token with the request, either as a cookie or a header. You can [learn more about Clerk's authentication tokens later in this guide](#stateful-authentication).

Reviewer 2: oops, I can fix that

1. The server responds to the browser's request by setting the session ID in a [`Set-Cookie`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) header in the response. This cookie will be automatically included in future requests from the browser in order to authenticate the user.
   <Video
     src="/docs/images/how-clerk-works/stateful-auth.mp4"
     width="1600"
     height="900"
     autoPlay
     muted
     loop
     playsInline
   />
1. The next time the browser sends a request to the server, it automatically includes the session cookie. The server checks the database for the session ID and retrieves the associated user ID and session metadata. If the session is valid and active, the server has verified that the user has authenticated, andcan then use the user ID to fetch any required user data and process the request.
   <Video
     src="/docs/images/how-clerk-works/stateful-auth-2.mp4"
     width="1600"
     height="900"
     autoPlay
     muted
     loop
     playsInline
   />

> [!NOTE]

> > If an attacker obtains your session ID, they can fully you as a user and take over your account. To mitigate this risk, Clerk implements the following security measures: 1. [HTTPS](https://developer.mozilla.org/en-US/docs/Glossary/HTTPS) is enforced to prevent [man-in-the-middle attacks](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) that could intercept session IDs in transit, and 2. the [`HttpOnly` flag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#httponly) is set on session cookies to prevent malicious JavaScript from accessing them through cross-site scripting (XSS) attacks. [Read more about Clerk's security measures](/docs/security/overview).

Author: This section needs to be heavily modified - we're not talking about Clerk's auth scheme here yet so an injection about how Clerk does auth feels out of place. Also, we don't actually use httponly for session cookies, so this isn't accurate. Will take another pass on this later but wanted to drop a note so i dont forget at least

Reviewer 2: oh, i wrote that incorrectly about the `httponly` flag :(
and i see, good point about injecting how clerk handles it! i thought we were mentioning things about other flows (in this case, stateful) in order to compare to clerk's flows. 
i would ask then, what is the purpose of providing this information? why provide this callout for users?

Furthermore, even after rotating keys, the revocation may be delayed if your application caches the public key used for verification - a common practice for performance optimization. The cached key would continue to validate the old tokens until the cache expires.

This limitation poses significant security risks, as it hampers your ability to quickly respond to security incidents that require immediate session termination for specific users.

## Clerk's authentication model

Clerk implements a hybrid authentication model that combines the benefits of both stateful and stateless approaches while mitigating their respective drawbacks, at the cost of adding a substantial amount of complexity to the implementation on Clerk's side. For a developer implementing Clerk, like you, this is all upside since the complexity is handled internally by Clerk. The hybrid model incorporates the same flow when signing in as the stateless flow, but with a few key innovations:

- A short-lived **session token** (60 seconds) combined with a background token refresh mechanism.

> - A **client token** to verify the user's authentication state, in conjunction with the session token.

Reviewer 2: let's pair and drive it home more!
